### ISSUE TITLE
Training Time

### ISSUE THREAD

**Issue by MaxLeung99** [yilundu/improved_contrastive_divergence#14]
Hi! Thanks for your great work! I'm running the code to reproduce the results. According to your paper,  it took you 2 days to train the model for CIFAR10. However, when I'm training the model on a single RTX 3090. I think the speed is relatively slow. So could you please share some exact configuration of your training? I'm training the model for 10000 epochs. And the speed, for now, is 4 minutes per epoch which is not reasonable.  So I guess you probably didn't train the model exactly for 10000 epochs and maybe stopped earlier?

**Comment by yilundu** [comment]
Hi Max, it's been a while since I've run this code -- IIRC we didn't actually train for that long -- only around 10k to 20k iterations.

**Comment by MaxLeung99** [comment]
Thank you for your quick reply. I will try to train the model for 10k to 20k and see how it works.

**Comment by MaxLeung99** [comment]
Sorry to bother you again. Could you please share the command used to generate samples? (CIFAR10 setting, this repo not igebm). And I find that when I use igebm ckpt(trained by myself) to sample images, sometimes some ckpt will fail to produce meaningful samples. Is it a normal situation?

**Comment by yilundu** [comment]
Sure I think you can run this command:

`python test_inception.py --exp cifar10_model --num_steps 10 --batch_size=512 --step_lr=100.0 --resume_iter=120000 --im_number=10000`

**Comment by MaxLeung99** [comment]
Sorry. A little bit of confusion here. So you train the model for 120k, not 10k~20k?

**Comment by yilundu** [comment]
I believe the results do not change much with more training -- so I think results with 10k-20k should be fairly comparable. I ran these experiments a long time ago -- but I guess at the time I must just let one model training for a very long time.

**Comment by MaxLeung99** [comment]
Ok. That's pretty clear for me now! Thanks for your patience and your answer!

**Comment by MaxLeung99** [comment]
Hi, I trained a model for 120k this week. I adjust the parameters by myself and the best result I can get is FID 55.529, IS 6.384. The command for the best result is python test_inception.py --exp cifar10_model --num_steps 20 --batch_size=512 --step_lr=100.0 --repeat_scale=30 --nomix=5 --resume_iter=120000 --im_number=50000. I'm not sure what causes the gap. I also tested the command you gave in another issue (nomix=4, step_lr=10.0) which also generate almost black output. So I want to ask if there is any missing point.

**Comment by yilundu** [comment]
Hi Max -- can you share the sampled images you got from the model?

**Comment by MaxLeung99** [comment]
![30fccf12-327c-42cc-b4ea-04a23a1c9003](https://user-images.githubusercontent.com/16420891/189982357-9596e283-cb7b-4099-ab28-cea15de5ebf6.png)
This is the best result so far.
![unnamed](https://user-images.githubusercontent.com/16420891/189983774-79f47557-788b-4987-a847-a647f3dbecf3.png)
If I use nomix=30, this is the result.  I think the images as you said in your paper lose their background and I think this is right. But I think we can see some images are crashed. Even when nomix=10, I can see some images like a mess and we cannot identify the object in it and it seems that the generation fails. But if I set nomix = 5, I think the image quality could be improved with longer Langevin dynamics. So nomix here may be important.

**Comment by yilundu** [comment]
Yeah the generations seem a bit strange to me -- can you send the tensorboard log of energy values over the course of training?

**Comment by MaxLeung99** [comment]
Sure. I actually changed the logger into wandb. But I think that would not affect the results. 
<img width="431" alt="image" src="https://user-images.githubusercontent.com/16420891/189988397-abf8520d-501b-401a-aa3d-900f2e9a3683.png">

<img width="429" alt="image" src="https://user-images.githubusercontent.com/16420891/189988468-fbe30b19-ab4c-49a0-877c-cb4d49a94043.png">
I think the pos energy fluctuates within a small range.

<img width="423" alt="image" src="https://user-images.githubusercontent.com/16420891/189988629-1ef8d68e-fcb5-4673-a851-b6bc83ad26e4.png">

<img width="431" alt="image" src="https://user-images.githubusercontent.com/16420891/189988692-23d8e080-fe7c-4504-9afe-9fd9472c2fb9.png">

It seems a little bit strange. But I used exactly the same command you listed for CIFAR10-uncond

**Comment by yilundu** [comment]
I see -- yes the code seems to have become unstable around step 1.5k -- that should likely get you the best samples. What happens if you try to evaluate the model at that timestep?

**Comment by MaxLeung99** [comment]
Sure. I'll give it a try and see what happens.

**Comment by MaxLeung99** [comment]
![image](https://user-images.githubusercontent.com/16420891/190066768-a36d0fc5-8d5f-433f-9abc-a312591525be.png)
The images are getting better. The FID is 51.661, and IS is 6.987. Do you use the combined model or ensemble during testing? And I find that in most experiments, the training becomes unstable after 50k iterations. Is this abnormal?

**Comment by yilundu** [comment]
We use a single model during testing -- yes this is normal -- around 50k iterations the energy landscape becomes too non-smooth for the sampler to find a good negative sample. You can partially circumvent this by increasing the number of sampling steps over the course of training. You could also try sampling a bit before it becomes unstable (maybe 10k or so iterations before). Finally, you can try reducing the number of sampling steps -- the image above still look a bit saturated.

**Comment by MaxLeung99** [comment]
Ok! I'll keep trying.

**Comment by MaxLeung99** [comment]
Hi ! When I was tuning the sampling with different parameters, I noticed that when I removed the transform or set nomix to a large number (30). The images are totally saturated. This happens even with short sampling steps. Maybe this results from the discrepancy between training and sampling (training always enables transform). 
![image](https://user-images.githubusercontent.com/16420891/190622990-254df61c-7d54-4317-976b-3e97a97d6b1b.png)
Have you encountered problems like this? I just want to figure out whether my training model is wrong or if it's a problem with sampling.

**Comment by yilundu** [comment]
Hi -- I also sometimes encountered this issue - my thought is that there is disconnect between training and test time sampling. I believe you can remedy this issue a bit if you train with slightly less data augmentation at training time -- and also decrease the step size of gradient descent (during training)
